### ISSUE TITLE
[Bash] highlight nested quotes and shell expansions

### ISSUE THREAD

**Issue by zufuliu** [ScintillaOrg/lexilla#154]
This is a big change, I would like to spit it as following small changes:
* `noexcept` and clang-tidy suggested changes.
* change `BASH_CMD_*` and `BASH_DELIM_*` macros into scoped enumeration.
* treat parameter expansion `${}` same as backticks on handling nested quotes and shell expansions, as [the document](https://www.gnu.org/software/bash/manual/bash.html#Shell-Parameter-Expansion) says:
> When braces are used, the matching ending brace is the first ‘}’ not escaped by a backslash or within a quoted string, and not within an embedded arithmetic expansion, command substitution, or parameter expansion. 
* Remove `QuoteCls Quote;` variable as single quoted string has no inner states.
* Move out `QuoteStackCls` definition, and change it to use `QuoteCls` array as stack.
* Change the for loop to while loop, it currently has bug on highlighting single punctuation [special parameters](https://www.gnu.org/software/bash/manual/bash.html#Special-Parameters). e.g. whole `$?1` highlighted as variable instead of just `$?`.
* highlight nested quotes and shell expansions for double quoted string, backticks and parameter expansion. `$()` command substitution still colored as `SCE_SH_BACKTICKS`.
* Simplify `case SCE_SH_HERE_Q` by using `sc.Match()` to match the delimiter.
* highlight shell expansions in here document.

**Comment by zufuliu** [comment]
First change, add `noexcept` and `const`, use `nullptr` and C++ include, remove unused destructor.
[bash-tidy.zip](https://github.com/ScintillaOrg/lexilla/files/11360346/bash-tidy.zip)

**Comment by nyamatongwe** [comment]
Looks good.

Changed order of includes to match `lexilla/scripts/HeaderOrder.txt`. The C++ versions are in a different order to the C includes but it would cause too much profitless code churn to fix.

**Comment by zufuliu** [comment]
[bash-enum.zip](https://github.com/ScintillaOrg/lexilla/files/11360646/bash-enum.zip)

scoped enumeration changes: `BASH_CMD_*` changed to `CmdState`, `BASH_DELIM_` change to `QuoteStyle` (will be style field for `QuoteCls`), renamed `BASH_DELIM_STACK_MAX` to `BASH_QUOTE_STACK_MAX`, added enumratin for `TestExprType`.

**Comment by nyamatongwe** [comment]
Added in `bash-enum.zip` but synchronized some comments with the new enum names. Also added some more test cases.

Started working on nested and multi-line tests but am not sure which tests are good. Here is the current state:
~~~bash
# Nested elements

# String with backtick inclusion
"x`ls`"
# Nested string
"x`ls "*.c"`"
# Not terminated at first "
"x`ls" # "`" #

# String with command inclusion
"x$(ls)"

# Nested command
$(ls -la$(ls *.c))

# Check strings and backticks in command
$('x' "x" `ls` $'x' $"x")

# $( not terminated by ) if contains unterminated string
$('x) # ') #
$("x) # ") #
$(`x) # `) #
$($'x) # ') #
$($"x) # ") #

# Multi-line
$(ls |
more)

$(
`x`
"x"
`ls`
$'x'
$"x"
)
#end -- checks termination of previous
~~~

**Comment by zufuliu** [comment]
There also needs a test for parameter expansion:
```bash
var=abcdef
sub=abc
rep='& '
echo ${var/$sub/"${rep}}"} #
```

Change `SCE_SH_PARAM` to be handled same as `SCE_SH_BACKTICKS`:
[bash-param.zip](https://github.com/ScintillaOrg/lexilla/files/11364132/bash-param.zip)

**Comment by zufuliu** [comment]
Remove `QuoteCls Quote;` variable.
[bash-single-quote.zip](https://github.com/ScintillaOrg/lexilla/files/11370521/bash-single-quote.zip)

**Comment by zufuliu** [comment]
Move out `class QuoteCls` and `class QuoteStackCls`, removed unused ` && QuoteStack.Up != '\\'` test.
[bash-quote-stack.zip](https://github.com/ScintillaOrg/lexilla/files/11372713/bash-quote-stack.zip)

**Comment by nyamatongwe** [comment]
`QuoteCls::Clear` is never called and `QuoteCls::Open` is only called from `QuoteCls::Start` so its code could be moved there.

**Comment by zufuliu** [comment]
Change for loop to while loop, new test case:
```bash
echo $?Status
```

[bash-while-loop.zip](https://github.com/ScintillaOrg/lexilla/files/11396386/bash-while-loop.zip)

**Comment by nyamatongwe** [comment]
This seems a bit heavy to just avoid the `sc.Forward()` for `$` and removes the for loop's reassurance of progression and thus termination.

The code could be changed to check that `sc.chNext` needs a pair like the following although that does state the set of paired characters twice. Alternatively use `sc.chNext` in the if block and `sc.Forward()` after.
~~~patch
@@ -856,6 +856,10 @@ void SCI_METHOD LexerBash::Lex(Sci_PositionU startPos, Sci_Position length, int
 					continue;
 				}
 				sc.SetState(SCE_SH_SCALAR);
+				if (!AnyOf(sc.chNext, '{', '\'', '\"', '(', '`')) {
+					// scalar has no delimiter pair
+					continue;
+				}
 				sc.Forward();
 				if (sc.ch == '{') {
 					sc.ChangeState(SCE_SH_PARAM);
@@ -872,8 +876,6 @@ void SCI_METHOD LexerBash::Lex(Sci_PositionU startPos, Sci_Position length, int
 				} else if (sc.ch == '`') {	// $` seen in a configure script, valid?
 					sc.ChangeState(SCE_SH_BACKTICKS);
 					QuoteStack.Start(sc.ch, QuoteStyle::Backtick);
-				} else {
-					continue;	// scalar has no delimiter pair
 				}
 			} else if (sc.Match('<', '<')) {
 				sc.SetState(SCE_SH_HERE_DELIM);
~~~

[continue.patch](https://github.com/ScintillaOrg/lexilla/files/11401980/continue.patch)

**Comment by zufuliu** [comment]
OK, seems better to move `AnyOf` as a function.

I think convert for loop to while is still needed, current code for handling nested highlighting:
https://github.com/zufuliu/notepad2/blob/main/scintilla/lexers/LexBash.cxx#L645
```cpp
case SCE_SH_STRING_DQ:	// delimited styles, can nest
case SCE_SH_PARAM: // ${parameter}
case SCE_SH_BACKTICKS:
	if (sc.ch == '\\') {
		sc.Forward();
	} else if (sc.ch == QuoteStack.Current.Down) {
		if (QuoteStack.CountDown(sc, cmdState)) {
			continue;
		}
	} else if (sc.ch == QuoteStack.Current.Up) {
		QuoteStack.Current.Count++;
	} else {
```

and `CountDown()` (so it need to avoid extra `sc.Forword()` by `continue`):
https://github.com/zufuliu/notepad2/blob/main/scintilla/lexers/LexBash.cxx#L259
```cpp
if (Current.Count == 0) {
	cmdState = CmdState::Body;
	const int outer = Current.Outer;
	if (Depth > 0) {
		Pop();
	} else {
		Clear();
	}
	sc.ForwardSetState(outer);
	return true;
}
```

**Comment by zufuliu** [comment]
Whole code block for `sc.ch == '$'` will be extracted as method `QuoteStack.Expand(sc, cmdState);`, I think `AnyOf(sc.chNext)` change can be delayed then (e.g. to avoid `continue` after `Expand()`)?
https://github.com/zufuliu/notepad2/blob/main/scintilla/lexers/LexBash.cxx#L782
```cpp
} else if (sc.ch == '$') {
	QuoteStack.Expand(sc, cmdState);
	continue;
```

**Comment by nyamatongwe** [comment]
> I think convert for loop to while is still needed, current code for handling nested highlighting:
> ... 
> and `CountDown()` (so it need to avoid extra `sc.Forword()` by `continue`): 

So the key concern here is that the terminating character for a possibly-nested element is styled with `ForwardSetState()` but that reveals the following character `sc.ch` which needs to be processed by the top-half (state terminating) part of the loop before another `Forward()` occurs?

**Comment by zufuliu** [comment]
Yes, consider `"${var}"`, after process `}`, `sc.ch` is the quote, control should transfer to `SCE_SH_STRING` case without forward.

**Comment by zufuliu** [comment]
Remove `$` (starts variable substitution or expansion) from `setParam` to fix another variable highlighting bug:
```bash
$$PID
$var${var}
```

```diff
diff --git a/lexers/LexBash.cxx b/lexers/LexBash.cxx
index def2b49d..e2a94371 100644
--- a/lexers/LexBash.cxx
+++ b/lexers/LexBash.cxx
@@ -363,7 +363,7 @@ void SCI_METHOD LexerBash::Lex(Sci_PositionU startPos, Sci_Position length, int
        setMetaCharacter.Add(0);
        const CharacterSet setBashOperator(CharacterSet::setNone, "^&%()-+=|{}[]:;>,*/<?!.~@");
        const CharacterSet setSingleCharOp(CharacterSet::setNone, "rwxoRWXOezsfdlpSbctugkTBMACahGLNn");
-       const CharacterSet setParam(CharacterSet::setAlphaNum, "$_");
+       const CharacterSet setParam(CharacterSet::setAlphaNum, "_");
        const CharacterSet setHereDoc(CharacterSet::setAlpha, "_\\-+!%*,./:?@[]^`{}~");
        const CharacterSet setHereDoc2(CharacterSet::setAlphaNum, "_-+!%*,./:=?@[]^`{}~");
        const CharacterSet setLeftShift(CharacterSet::setDigits, "$");
```

**Comment by zufuliu** [comment]
1. Move `cmdDelimiter`, `bashStruct` and `bashStruct_in` into the class as they are expansive to setup.
2. remove `WordList wlNew;`.
3. replace `styler.GetLine(sc.currentPos)` with `sc.currentLine`.


[bash-WordList-currentLine.patch](https://github.com/ScintillaOrg/lexilla/files/11451483/bash-WordList-currentLine.patch)

**Comment by zufuliu** [comment]
Implement nested highlighting, new test case:
```bash
echo "$((1 + 2))" #
echo "$[1 + 2]" #
```

[bash-nested.patch](https://github.com/ScintillaOrg/lexilla/files/11463428/bash-nested.patch)

**Comment by nyamatongwe** [comment]
The `bash-nested.patch` diff implements the feature well for all the examples I have tried.

However, I don't think this feature will be universally wanted, including by myself and the original lexer author Kein-Hong Man. For me, its highlighting smaller elements that I am less interested in most of the time, similar to highlighting escape sequences in C++ strings which I turn off.

This new behaviour should be a choice although it may not be just a binary choice - some may want more fine-grained styling inside `$()` commands but not inside `""` strings.

**Comment by zufuliu** [comment]
OK, then need to add property for each nested quote style + scalar variable, and what would the property names?

**Comment by nyamatongwe** [comment]
I'm unsure whether or not the multiple settings is required but I'm fairly sure at least one is. It could start with a single option and grow to a full set when there is demand or it seems worthwhile.

Perhaps a single `lexer.bash.styling.within.quotes` to match `lexer.baan.styling.within.preprocessor`. Or `lexer.bash.styling.inside.quotes`.

If there are multiple settings they should match the class name or the element name in bash documentation
`SCE_SH_BACKTICKS` -> `lexer.bash.styling.within.backticks`, or
`SCE_SH_STRING` -> `lexer.bash.styling.inside.string`.

**Comment by zufuliu** [comment]
`styling.within` or `styling.inside` seems suitable for #153, main change for this is highlighting [Shell Expansions](https://www.gnu.org/software/bash/manual/bash.html#Shell-Expansions)([Shell Parameter Expansion](https://www.gnu.org/software/bash/manual/bash.html#Shell-Parameter-Expansion), [Command Substitution](https://www.gnu.org/software/bash/manual/bash.html#Command-Substitution) and [Arithmetic Expansion](https://www.gnu.org/software/bash/manual/bash.html#Arithmetic-Expansion)) inside string, backticks/command substitution, parameter expansion.

**Comment by zufuliu** [comment]
[bash-nested-prop.patch](https://github.com/ScintillaOrg/lexilla/files/11470827/bash-nested-prop.patch)

Added `lexer.bash.styling.inside.[string | backticks | parameter]`.

**Comment by nyamatongwe** [comment]
Committed with an added `[[nodiscard]]` for `stylingInside`.

Duplicated `Nested.bsh` as `NestedStyledInside.bsh` to test `lexer.bash.styling.inside.*`.

**Comment by zufuliu** [comment]
[bash-heredoc.patch](https://github.com/ScintillaOrg/lexilla/files/11476964/bash-heredoc.patch)

Use `sc.Match()` to match here document delimiter.

**Comment by zufuliu** [comment]
[bash-heredoc3.patch](https://github.com/ScintillaOrg/lexilla/files/11477957/bash-heredoc3.patch)

Simplify preview patch.

**Comment by zufuliu** [comment]
[bash-heredoc-nested.patch](https://github.com/ScintillaOrg/lexilla/files/11487030/bash-heredoc-nested.patch)

Implement nested highlighting inside here document, new test case (undocumented at https://www.gnu.org/software/bash/manual/bash.html#Here-Documents, but same behavior as Perl when there are backslash inside delimiter):
```bash
cat <<-\EOF
	$((1+2))
	`date`
	$(date)
EOF
```

**Comment by zufuliu** [comment]
[bash-heredoc-nested2.patch](https://github.com/ScintillaOrg/lexilla/files/11492507/bash-heredoc-nested2.patch)

Previous patch has bugs when `stylingInsideHeredoc` is not set (no quote counting), updated the patch, how ever it's only works when `stylingInsideHeredoc` is enabled for following code:
```bash
cat <<EOF
heredoc line 1
$(echo \
EOF
)
heredoc line 3
EOF

echo $(cat <<'EOF'
heredoc line 1
$(echo \
EOF
)
#

```

**Comment by nyamatongwe** [comment]
> new test case (undocumented ...), but same behavior as Perl when there are backslash inside delimiter

If its undocumented, where is the need coming from?

**Comment by zufuliu** [comment]
> where is the need coming from?

it's common in configure script, e.g. there 20 `<<\` in https://github.com/eblot/newlib-cygwin/blob/master/configure

Add `QuoteStyle::HereDoc` to fix first failure when `stylingInsideHeredoc` not enabled.
[bash-heredoc-nested3.patch](https://github.com/ScintillaOrg/lexilla/files/11492596/bash-heredoc-nested3.patch)

**Comment by zufuliu** [comment]
Moved `HereDoc` enumerator to bellow `LString`.
[bash-heredoc-nested4.patch](https://github.com/ScintillaOrg/lexilla/files/11492649/bash-heredoc-nested4.patch)

**Comment by zufuliu** [comment]
[bash-heredoc-nested5.patch](https://github.com/ScintillaOrg/lexilla/files/11492753/bash-heredoc-nested5.patch)

Change the code to use `QuoteStack.Start(-1, QuoteStyle::HereDoc, SCE_SH_DEFAULT);` to avoid counting.
Fix second failure requires highlighting inside command substitution.

**Comment by nyamatongwe** [comment]
With this `bash-heredoc-nested5.patch` change, here-docs style the following byte. With Windows `\r\n` line ends this styles just the `\r` as `SCE_SH_HERE_Q`.

It really is worthwhile running the tests after making a change.
~~~
Lexing bash\x.bsh
C:\u\hg\lexilla\test\examples\bash\x.bsh:21: different styles between \r and \n at 769: 13, 0
~~~

`HereDocCls::Append` should not be `noexcept` as it can fail producing undefined behaviour if the fixed-length buffer is exhausted. The caller protects against this occurring but its a logically fallible action and may fail (in different ways) if `Delimiter` was changed to a `std::string` for example.

**Comment by zufuliu** [comment]
[bash-heredoc-nested6.patch](https://github.com/ScintillaOrg/lexilla/files/11506955/bash-heredoc-nested6.patch)

test failure caused by `sc.ForwardSetState(outer);`.

**Comment by zufuliu** [comment]
Another fix to not do the magic call `CountDown()` for `SCE_SH_HERE_Q`.
[bash-heredoc-nested7.patch](https://github.com/ScintillaOrg/lexilla/files/11512143/bash-heredoc-nested7.patch)

**Comment by nyamatongwe** [comment]
Also ensured initialization to avoid warnings with d56956e060a50d0efdd46fa429335547a55b4c5f.

**Comment by nyamatongwe** [comment]
Included in release 5.2.5.
